### PR TITLE
Allow writing out raw JSON.

### PR DESCRIPTION
**Pull request by NightlyNexus**
taking a shot at #65.

### CODE REVIEW

File: moshi/src/test/java/com/squareup/moshi/JsonWriterTest.java [hunk 1]
```diff
@@ -571,4 +572,15 @@
...
+
+  @Test public void valueFromSource() throws IOException {
+    JsonWriter writer = factory.newWriter();
+    writer.beginObject();
+    writer.name("a");
+    writer.value(new Buffer().writeUtf8("[\"value\"]"));
+    writer.name("c");
+    writer.value(3);
+    writer.endObject();
+    assertThat(factory.json()).isEqualTo("{\"a\":[\"value\"],\"c\":3}");
```
Comment (NightlyNexus) [inline moshi/src/test/java/com/squareup/moshi/JsonWriterTest.java]: might delete this. covered in the separated tests.

File: moshi/src/main/java/com/squareup/moshi/JsonUtf8Writer.java [hunk 2]
```diff
@@ -244,6 +245,24 @@ private void writeDeferredName() throws IOException {
     return this;
   }
 
+  @Override public JsonWriter value(BufferedSource source) throws IOException {
+    if (promoteValueToName) {
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonUtf8Writer.java]: I don’t think we should support this.

Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonUtf8Writer.java]: ... instead we should probably just fail. It’s weird.

File: moshi/src/main/java/com/squareup/moshi/JsonValueWriter.java [hunk 3]
```diff
@@ -182,6 +183,28 @@ public Object root() {
...
+  @Override public JsonWriter value(BufferedSource source) throws IOException {
+    Object value = JsonReader.of(source).readJsonValue();
+    if (value instanceof String) {
+      return value((String) value);
+    }
+    if (value instanceof Number) {
+      return value((Number) value);
+    }
+    if (promoteValueToName) {
+      throw new AssertionError("This is not a map key.");
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonValueWriter.java]: same

Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonValueWriter.java]: Should be IllegalStateException

File: moshi/src/main/java/com/squareup/moshi/JsonValueWriter.java [hunk 4]
```diff
@@ -182,6 +183,28 @@ public Object root() {
...
+    }
+    boolean serializeNulls = this.serializeNulls;
+    this.serializeNulls = true;
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonValueWriter.java]: nice

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.java [hunk 5]
```diff
@@ -335,6 +336,8 @@ public final void setSerializeNulls(boolean serializeNulls) {
    */
   public abstract JsonWriter value(@Nullable Number value) throws IOException;
 
+  public abstract JsonWriter value(BufferedSource source) throws IOException;
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonWriter.java]: This is a clever API. Nice.

File: moshi/src/test/java/com/squareup/moshi/JsonUtf8WriterTest.java [hunk 6]
```diff
@@ -107,4 +107,20 @@
     // JsonWriter doesn't attempt to detect duplicate names
     assertThat(buffer.readUtf8()).isEqualTo("{\"a\":1,\"a\":2}");
   }
+
+  @Test public void valueFromSource() throws IOException {
```
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/JsonUtf8WriterTest.java]: nice

File: moshi/src/test/java/com/squareup/moshi/JsonValueWriterTest.java [hunk 7]
```diff
@@ -239,6 +241,22 @@
     assertThat((List<?>) writer.root()).isEqualTo(numbers);
   }
 
+  @Test public void valueFromSource() throws IOException {
```
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/JsonValueWriterTest.java]: awesome

File: moshi/src/test/java/com/squareup/moshi/PromoteNameToValueTest.java [hunk 8]
```diff
@@ -333,4 +335,19 @@
...
+    writer.promoteValueToName();
+    writer.value(new Buffer().writeUtf8("\"a\""));
+    writer.value("a value");
+    writer.promoteValueToName();
+    writer.value(new Buffer().writeUtf8("2"));
+    writer.value("2 value");
+    writer.name("c");
+    writer.value(3);
+    writer.endObject();
+    assertThat(factory.json()).isEqualTo("{\"a\":\"a value\",\"2.0\":\"2 value\",\"c\":3}");
```
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/PromoteNameToValueTest.java]: the `2.0` really suggests to me that promoteNameToValue is not a good fit

File: moshi/src/main/java/com/squareup/moshi/JsonUtf8Writer.java (hunk 2 above)
Comment (NightlyNexus) [reply moshi/src/main/java/com/squareup/moshi/JsonUtf8Writer.java]: Agree! Hurts my head to think about this usage. IllegalStateException is best.

File: moshi/src/main/java/com/squareup/moshi/JsonValueWriter.java (hunk 4 above)
Comment (NightlyNexus) [reply moshi/src/main/java/com/squareup/moshi/JsonValueWriter.java]: I might document this explicitly in the Javadoc.

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.java (hunk 5 above)
Comment (NightlyNexus) [reply moshi/src/main/java/com/squareup/moshi/JsonWriter.java]: I'll write up a Javadoc since the API looks good to you!

Comment (NightlyNexus) [reply moshi/src/main/java/com/squareup/moshi/JsonWriter.java]: I considered valueRaw(String) as a convenience API for Buffer.writeUtf8(), but I'm hoping this is used much more for appending from non-memory sources.

Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonWriter.java]: Yeah, offering anything that takes a String is probably a disservice; users will call that thinking it’s the same as `value(String)`

**Comment by NightlyNexus** [comment]
added a Javadoc. I'm block this PR on that PR that adds the error message for invalid map keys.

**Pull request merged**